src/Paprika/Store/DataPage.cs

```diff
@@ -129,187 +90,86 @@ private static void Set(DbAddress at, in NibblePath key, in ReadOnlySpan<byte> d
+            {
+                childAddr = payload.Buckets[GetIndex(k)];
+                if (childAddr.IsNull == false && batch.WasWritten(childAddr))
                 {
+                    // Delete the k in this page just to ensure that the write-through will write the last value.
                     map.Delete(k);
-                    break;
-                }
 
-                // Try set in the map
-                if (map.TrySet(k, data))
-                {
-                    // Update happened, return
-                    break;
+                    k = k.SliceFrom(ConsumedNibbles);
+                    current = childAddr;
+                    continue;
                 }
+            }
 
-                // Failed to set, use overflow.
-                // 1. check if delete, then delete in both.
-                // 2. flush some down
-                // 3. retry set
-                var overflow = GetWritableOverflow(batch, ref payload);
+            // Try to write in the map
+            if (map.TrySet(k, data))
+            {
+                // Update happened, return
+                break;
+            }
 
-                // Assert existence
-                Debug.Assert(payload.Buckets[LeafMode.Bucket0].IsNull == false);
-                Debug.Assert(payload.Buckets[LeafMode.Bucket1].IsNull == false);
+            // First, try to flush the existing
+            if (TryFindMostFrequentExistingNibble(oddity, map, payload.Buckets, out var nibble))
+            {
+                childAddr = EnsureExistingChildWritable(batch, ref payload, nibble);
+                FlushDown(map, nibble, childAddr, batch);
 
-                // Asserting written status
-                Debug.Assert(batch.WasWritten(payload.Buckets[LeafMode.Bucket0]));
-                Debug.Assert(batch.WasWritten(payload.Buckets[LeafMode.Bucket1]));
+                // Spin one more time
+                continue;
+            }
 
-                if (data.IsEmpty)
-                {
-                    map.Delete(k);
-                    overflow.Delete(k);
-                    break;
-                }
+            // None of the existing was flushable, find the most frequent one
+            nibble = FindMostFrequentNibble(oddity, map);
 
-                // Move down
-                map.MoveNonEmptyKeysTo(overflow.AsSource(), treatEmptyAsTombstone: true);
+            // Ensure that the child page exists
+            childAddr = payload.Buckets[nibble];
+            Debug.Assert(childAddr.IsNull,
+                "Address should be null. If it wasn't it should be the case that it's found above");
 
-                // Try set again
-                if (map.TrySet(k, data))
-                {
-                    // Update happened, return
-                    break;
-                }
+            // Create a child
+            var lvl = (byte)(page.Header.Level + ConsumedNibbles);
+            batch.GetNewPage<BottomPage>(out childAddr, lvl);
```

Comment:
I might be completely wrong here (or it might be already existing and I don't see it :P) : While allocating a new BottomPage, would it be useful to do some form of bulk page allocation? Most BottomPages would eventually need to allocate its children and having the children pages contiguous in the memory would make most operations faster?

Reply:
> Most BottomPages would eventually need to allocate its children.

They might have some children depending on the data but don't have to. We could allocate more of them but this would allocate empty space and we don't want to over allocate. With the current approach, we allocate only when there's something to flush down to a given page. This ensures that we save up a lot.

> having children pages contiguous in the memory would make most operations faster

Paprika uses `COW`. What not is contiguous, probably will not be after a few COWs. Imagine the root being COWed in one batch, then the child in next. A probability that they will be _contiguous_ is quite unlikely I believe. One thing to mention is that we do sort abandoned pages and serve them that way. We also allocate from the end when no abandoned can be reused. This should increase the probability of pages allocated one after another and as we COW as we walk through the pages... maybe we get a higher chance of continuity for free?